pkg/cli/image/strategy/simplelookup.go

```diff
@@ -0,0 +1,166 @@
+// gathered from ImageContentSourcePolicy objects and user-passed image.
+// Will lookup from cluster or from ImageContentSourcePolicy file passed from user.
+// Image reference of user-given image may be different from original in case of mirrored images.
+func (s *simpleLookupICSP) resolve(ctx context.Context, imageRef reference.DockerImageReference) error {
+	if s.alternates != nil {
+		return nil
+	}
+
+	var icspList []operatorv1alpha1.ImageContentSourcePolicy
+	var err error
+	if len(s.icspFile) > 0 {
+		klog.V(5).Infof("Reading ICSP from file %s", s.icspFile)
+		icspList, err = s.addICSPsFromFile()
+		if err != nil {
+			return err
+		}
+	} else {
+		// log errors from accessing cluster since this function has no way of knowing whether it should or should not succeed
+		klog.V(5).Infof("Reading ICSP from cluster")
+		icspList, err = s.addICSPsFromCluster(ctx)
+		if err != nil {
+			klog.V(4).Infof("No alternative image sources gathered from cluster: %v", err)
+		}
+	}
+	imageRefList, err := s.alternativeImageSources(imageRef, icspList)
+	if err != nil {
+		return err
+	}
+	s.alternates = imageRefList
+	return nil
+}
+
+// alternativeImageSources returns unique list of DockerImageReference objects from list of ImageContentSourcePolicy objects
+func (s *simpleLookupICSP) alternativeImageSources(imageRef reference.DockerImageReference, icspList []operatorv1alpha1.ImageContentSourcePolicy) ([]reference.DockerImageReference, error) {
+	var imageSources []reference.DockerImageReference
+	// make sure at least 1 imagesource
+	// ie, make sure the image passed is included in image sources
+	// this is so the user-given image ref will be tried
+	imageSources = append(imageSources, imageRef.AsRepository())
+	klog.V(5).Infof("%v ImageReference added to potential ImageSourcePrefixes from ImageContentSourcePolicy", imageRef.AsRepository())
+	for _, icsp := range icspList {
+		repoDigestMirrors := icsp.Spec.RepositoryDigestMirrors
+		for _, rdm := range repoDigestMirrors {
+			var err error
+			rdmSourceRef, err := reference.Parse(rdm.Source)
+			if err != nil {
+				return nil, err
+			}
+			if imageRef.AsRepository() != rdmSourceRef.AsRepository() {
+				continue
+			}
+			klog.V(5).Infof("%v RepositoryDigestMirrors source matches given image", imageRef.AsRepository())
+			for _, m := range rdm.Mirrors {
+				mRef, err := reference.Parse(m)
+				if err != nil {
+					return nil, err
+				}
+				imageSources = append(imageSources, mRef)
+				klog.V(5).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageContentSourcePolicy", m)
+			}
+		}
+	}
+	uniqueMirrors := make([]reference.DockerImageReference, 0, len(imageSources))
+	uniqueMap := make(map[reference.DockerImageReference]bool)
+	for _, imageSourceMirror := range imageSources {
+		if _, ok := uniqueMap[imageSourceMirror]; !ok {
+			uniqueMap[imageSourceMirror] = true
+			uniqueMirrors = append(uniqueMirrors, imageSourceMirror)
+		}
+	}
```

Comment:
This does not follow https://github.com/openshift/api/blob/a99ffa1cac6709edf8f502b16890b16f9a557e00/operator/v1alpha1/types_image_content_source_policy.go#L34-L39 .

If at all possible, this should share the implementation in https://github.com/openshift/runtime-utils/blob/master/pkg/registries/registries.go (used e.g. in machine-config-* to generate CRI-O configuration from ICSP), and perhaps even the consumer from https://github.com/containers/image/blob/97b3ffa7bb92a7778238b58010454031b0c2cbee/pkg/sysregistriesv2/system_registries_v2.go#L815 + https://github.com/containers/image/blob/97b3ffa7bb92a7778238b58010454031b0c2cbee/pkg/sysregistriesv2/system_registries_v2.go#L130 , OTOH that would require API additions to make it usable with in-memory configs) instead of maintaining an independent implementation. There are RFEs to add hostname wildcards and the like to the configuration; of course it’s possible to maintain independent implementations but there are enough corner cases that sharing an implementation is the easiest way to remain consistent.

---

(I haven’t read the code in detail, but the behavior of ICSP is not “fallback alternatives”; the mirrors, if any are tried _before_ the primary location, e.g. so that in a disconnected environment there isn’t a firewall/DNS noise, and delays, for attempts to reach the global internet for images that are available on mirrors.)

Reply:
this is intentional, oc is not a cluster.  For cli commands not tied to the behavior of a cluster the appropriate behavior is fallback.  Not the least of which is that ICSP defines alternatives,  and the tradeoffs different clients make drives different benefits.  A cli is low latency on the happy path.

Reply:
(To be explicit, my _primary_ concern is about sharing the implementation: the primary vs. mirrors ordering is far less important to me.)

If the user wants to hit the primary location directly, it’s trivial not to submit an ICSP to `oc`; all of this is only invoked on explicit user action to involve ICSP.

The linked bug also motivates this with a situation where the primary location is not available, and even suggests that `oc extract` is used during cluster installation. Both of these would benefit from using the mirrors first.

The `--cluster-icsp` option documentation also says:
> honor the ordering of those sources

Reply:
I expect a user to provide `--use-cluster-icsp` in scripting, and for it not to fail if one exists, and for us to "do the right thing for the user".  Since ICSP lookup on some commands (anything not doing bulk retrieval) is better suited for latency, I think we're reserving the right to "do the right thing" by underspecifying.

The clause `honor the ordering` should be removed from the flag.

Reply:
> The clause `honor the ordering` should be removed from the flag.

Removed